[#154957353] Support initial version and content

bandesz opened the pull request:
## What

### Initial state

If no resource versions exist you can set up this resource to emit an initial version with a specified content. This won't create a real resource in S3 but only create an initial version for Concourse. The resource file will be created as usual when you `get` a resource with an initial version.

You can define one of the following two options:

* `initial_path`: *Optional.* Must be used with the `regexp` option. You should set this to the file path containing the initial version which would match the given regexp. E.g. if `regexp` is `file/build-(.*).zip`, then `initial_path` might be `file/build-0.0.0.zip`. The resource version will be `0.0.0` in this case.

* `initial_version`: *Optional.* Must be used with the `versioned_file` option. This will be the resource version.

By default the resource file will be created with no content when `get` runs. You can set the content by using one of the following options:

* `initial_content_text`: *Optional.* Initial content as a string.

* `initial_content_binary`: *Optional.* You can pass binary content as a base64 encoded string.

## How to review

* Code review.
* Run the integration tests. For that you need to create two S3 buckets in dev: a version and an unversioned one. E.g. "${DEPLOY_ENV}-resource-test-versioned" and "${DEPLOY_ENV}-resource-test-unversioned"

```
docker build --target resource . -t s3-resource --build-arg \
  S3_TESTING_ACCESS_KEY_ID="${AWS_ACCESS_KEY_ID}" --build-arg \
  S3_TESTING_SECRET_ACCESS_KEY="${AWS_SECRET_ACCESS_KEY}" --build-arg \
  S3_TESTING_SESSION_TOKEN="${AWS_SESSION_TOKEN}" --build-arg \
  S3_TESTING_BUCKET="${DEPLOY_ENV}-s3-resource-test-unversioned" --build-arg \
  S3_VERSIONED_TESTING_BUCKET="${DEPLOY_ENV}-s3-resource-test-versioned" --build-arg \
  S3_TESTING_REGION="eu-west-1" --build-arg S3_ENDPOINT="https://s3.eu-west-1.amazonaws.com"
```

**Post-merge:** Once merged, this will be built as a container by DockerHub: https://hub.docker.com/r/governmentpaas/s3-resource/. Get the new tag from that, and update the various `[TMP]`/`[WIP]` commits in https://github.com/alphagov/paas-bootstrap/pull/157 https://github.com/alphagov/paas-release-ci/pull/60 https://github.com/alphagov/paas-cf/pull/1326 to use that revision of the container instead of the `initial_version` branch.

## Who can review

Not @46bit or me.

bandesz committed:
Add custom build hook to not run the integration tests on Docker Hub

bandesz committed:
Add documentation for S3 resource initialisation

alext commented on in/command.go:
```diff
@@ -146,6 +171,20 @@ func (command *Command) downloadFile(bucketName string, remotePath string, versi
+	if err != nil {
+		return err
+	}
+	defer localFile.Close()
```
By deferring the call to `Close()` like this, we're not checking the error. When writing files, the OS will often buffer writes until the file is closed, meaning that any errors arising from the actually write aren't returned until the close call. See https://joeshaw.org/dont-defer-close-on-writable-files/ for a more detailed explanation of this.

46bit replied:
Thanks Alex, that's very helpful to know.

alext commented on in/command_test.go:
```diff
@@ -372,6 +372,167 @@ var _ = Describe("In Command", func() {
+
+			It("it creates a file containing the initial text content", func() {
+				_, err := command.Run(destDir, request)
+				Ω(err).ShouldNot(HaveOccurred())
```
The existing tests in this file use the `Expect(foo).To(Bar())` style of gomega assertions. We should keep to that style for consistency.

Actually now that I read more closely, I see that there's a complete mixture of the 2 styles in both this file, and the repo. Joy. I'm not sure what the best thing to do is here...

46bit replied:
😞 Indeed, I chose not to point it out when pairing for that reason.

Commit:
Add ability to initialise non-existent resources

As described in https://github.com/concourse/s3-resource/issues/21, it
would be helpful if S3 resources could be configured to be available
even if they do not yet exist in S3. This commit adds parameters to
support that.

For S3 versioning, the `initial_version` parameter describes what the
version available should be named if the file does not exist in your
S3 bucket.

For filepath-based versioning, the `initial_path` parameter describes
the path to use if there are no regex matches in your S3 bucket.

A missing S3 resource will be empty by default, but you can specify
content for it. `initial_content_text` allows specifying text content,
or base64-encoded content can be provided to `initial_content_binary`.

Commit:
Add integration tests for resource initialisation
